### PR TITLE
Fix scaled softmax nonfinite semantics

### LINKED ISSUE #161150: multiplication and softmax throws incorrect results under torch.compile
### 🐛 Describe the bug

Maybe this issue is caused by the optimization related to FMA.

```python
import torch
torch.manual_seed(100)


class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.scale = torch.tensor([-1.7e14])

    def forward(self, x):
        for _ in range(3):
            x = x * self.scale
        softmaxed = torch.softmax(x, dim=0)
        return x, softmaxed


m = M()
opt = torch.compile(m)
inp = torch.randn((1,10))

m_out = m(inp)
opt_out = opt(inp)

print(f"[Eager] Softmax Input: ", m_out[0])
print(f"[Eager] Softmax output: ", m_out[-1])

print(f"[Compile] Softmax Input: ", opt_out[0])
print(f"[Compile] Softmax output: ", opt_out[-1])
```

```
[Eager] Softmax Input:  tensor([[-inf, inf, inf, -inf, inf, inf, inf, inf, -inf, inf]])
[Eager] Softmax output:  tensor([[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]])
[Compile] Softmax Input:  tensor([[-inf, inf, inf, -inf, inf, inf, inf, inf, -inf, inf]])
[Compile] Softmax output:  tensor([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])
```

Even though the softmax's input is not normal, the compiled model's softmax output is normal but an incorrect result.

### Versions

[pip3] numpy==2.0.2
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.3
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] torch==2.8.0
[pip3] triton==3.4.0
[conda] numpy                     2.0.2                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.8.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.8.90                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.8.93                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.8.90                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.10.2.21                pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.3.83                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.9.90                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.3.90                pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.8.93                pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.7.1                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.27.3                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.8.93                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.8.90                  pypi_0    pypi
[conda] torch                     2.8.0                    pypi_0    pypi
[conda] triton                    3.4.0                    pypi_0    pypi

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184046

Preserve the original scaled subtraction for nonfinite scaled-softmax inputs so Inductor does not turn eager NaNs into finite softmax results.

Fixes #161150

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: torch/_inductor/fx_passes/joint_graph.py [hunk 1]
```diff
@@ -1031,7 +1044,9 @@ def repl(inp, other):
 
 def div_softmax_pattern(match: Match, *, inp, other, dim, keepdim, dtype=None):
     def repl(inp, other):
+        scaled = inp / other
```
Comment (Skylion007) [inline torch/_inductor/fx_passes/joint_graph.py:1047]: You technically calculate (1 / other) and reuse it in the stable call now. You could multiply by the reciprocal in both places if you want and save some cycles.

**Review by Skylion007 — APPROVED** [review]

File: torch/_inductor/fx_passes/joint_graph.py [hunk 2]
```diff
@@ -1044,7 +1059,10 @@ def repl(inp, other):
         inp = inp * sign
         max_ = torch.amax(inp, dim=dim, keepdim=keepdim)
 
-        return (inp - max_) / (sign * other)
+        stable = (inp - max_) / (sign * other)
+        return _preserve_scaled_softmax_nonfinite_semantics(
```
Comment (frgossen) [inline torch/_inductor/fx_passes/joint_graph.py:1063]: Does this extra step have performance implications?

**Review by frgossen — APPROVED** [review]
Approving with comment.

File: torch/_inductor/fx_passes/joint_graph.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/fx_passes/joint_graph.py:1047]: My agent says good point. I am leaving this as-is for this correctness fix since the diff is approved and CI is green; if benchmarks flag this path, using a reciprocal in both branches is a reasonable follow-up.

File: torch/_inductor/fx_passes/joint_graph.py (hunk 2 above)
Comment (jansel) [reply torch/_inductor/fx_passes/joint_graph.py:1063]: My agent says yes, this adds finite checks and computes the original expression, so it has overhead on finite inputs. The tradeoff is preserving eager semantics when scaled values are nonfinite; static elision for provably safe scales can be a follow-up optimization.
